File: constrained-SOONO/data/eunite2001.m
min = 464; % smallest max-load in 1997-1998
max = 876; % largest max-load in 1997-1998

[y,x] = libsvmread('eunite2001');
m = svmtrain(y, x, '-s 3 -c 4096 -g 0.0625 -p 0.5');

[ty, tx] = libsvmread('eunite2001.t');

p = zeros(31,1)
for i=1:31,
  if i==1,
    txi = tx(i,:);
  else
    txi = [tx(i,1:9) (p(i-1)-min)/(max-min) tx(i-1,10:15)];
  end
  p(i) = svmpredict(ty(i), txi, m);
end

mape = 100/31*sum(abs((p-ty)./ty))
mse = (p-ty)'*(p-ty)/31
plot((1:31)', p, '--', (1:31)', ty, '-');
legend('predicted', 'real');
set(gca, 'fontsize', 18) ; 
set(findobj('Type', 'line'), 'LineWidth', 3)  

File: constrained-SOONO/data/eunite2001.py
import numpy as np
from sklearn.svm import SVR
from sklearn.datasets import load_svmlight_file
import matplotlib.pyplot as plt

# Initialization
min_val = 464  # smallest max-load in 1997-1998
max_val = 876  # largest max-load in 1997-1998

# Reading Data
x, y = load_svmlight_file(f='/home/willyzz/Documents/ocotools/constrained-SOONO/data/eunite2001')
x = x.toarray()
# Training the SVM Model
model = SVR(C=4096, gamma=0.0625, epsilon=0.5)
model.fit(x, y)

# Reading Test Data
tx, ty = load_svmlight_file(f='/home/willyzz/Documents/ocotools/constrained-SOONO/data/eunite2001.t')

# Prediction Loop
p = np.zeros(31)
for i in range(31):
    if i == 0:
        txi = tx[i, :]
    else:
        txi = np.concatenate([tx[i, :9], [(p[i-1] - min_val) / (max_val - min_val)], tx[i-1, 9:]])
    p[i] = model.predict([txi])[0]

# Calculating Metrics
mape = 100 / 31 * np.sum(np.abs((p - ty) / ty))
mse = np.mean((p - ty) ** 2)

print(f'MAPE: {mape}')
print(f'MSE: {mse}')

# Plotting Results
plt.plot(range(1, 32), p, '--', label='predicted')
plt.plot(range(1, 32), ty, '-', label='real')
plt.legend()
plt.xlabel('Day')
plt.ylabel('Load')
plt.title('Predicted vs Real Load')
plt.grid(True)
plt.show()